core-js/src/main/javascript/cdf/components/TextInputComponent.js

```diff
@@ -20,39 +20,72 @@ define([
   return BaseComponent.extend({
     update: function() {
       var myself = this;
-      var name = myself.name;
 
-      var selectHTML = "<input type='text' id='" + name + "' name='"  + name +
-        "' value='" + myself.dashboard.getParameterValue(myself.parameter) +
-        (myself.size ? ("' size='" + myself.size) : (myself.charWidth ? ("' size='" + myself.charWidth) : "")) +
-        (myself.maxLength ? ("' maxlength='" + myself.maxLength) : (myself.maxChars ? ("' maxlength='" + myself.maxChars) : "")) + "'>";
-      if(myself.size) {
-        Logger.warn("Attribute 'size' is deprecated");
-      }
-      if(myself.maxLength) {
-        Logger.warn("Attribute 'maxLength' is deprecated");
-      }
-
-      myself.placeholder().html(selectHTML);
+      myself._addHtmlToPlaceholder();
 
-      var el = $("#" + name);
+      var el = $("#" + myself.name);
 
       el.change(function() {
-        if(myself.dashboard.getParameterValue(myself.parameter) !== el.val()) {
-          myself.dashboard.processChange(name);
+        if (myself._isValueChanged(el)) {
+          myself._runValidation(el.val());
+          myself.dashboard.processChange(myself.name);
         }
       }).keyup(function(ev) {
-        if(ev.keyCode == 13 && myself.dashboard.getParameterValue(myself.parameter) !== el.val()) {
-          myself.dashboard.processChange(name);
+        if ((myself.refreshOnEveryKeyUp || ev.keyCode === 13) && myself._isValueChanged(el)) {
+          myself._runValidation(el.val());
+          myself.dashboard.processChange(myself.name);
         }
       });
 
+      if (myself.addClearIcon) {
+        $("#" + myself.name + "-clear-icon").click(function() {
+          el.val("");
+          myself.dashboard.processChange(myself.name);
+        });
+      }
+
       myself._doAutoFocus();
     },
 
     getValue: function() {
       return $("#" + this.name).val();
+    },
+
+    _addHtmlToPlaceholder: function() {
+      var componentHTML = "<input" +
+        " type='text'" +
+        " id='" + this.name + "'" +
+        " name='" + this.name + "'" +
+        " value='" + this.dashboard.getParameterValue(this.parameter) +
+        (this.size ? ("' size='" + this.size) : (this.charWidth ? ("' size='" + this.charWidth) : "")) +
+        (this.maxLength ? ("' maxlength='" + this.maxLength) : (this.maxChars ? ("' maxlength='" + this.maxChars) : "")) + "'>";
+      if (this.addClearIcon) {
+        var className = "clear-icon" + (this.clearIconClassName ? " " + this.clearIconClassName : "");
+        componentHTML += "<div id='" + this.name + "-clear-icon' class='" + className + "'></div>";
+      }
+      if (this.size) {
+        Logger.warn("Attribute 'size' is deprecated");
+      }
+      if (this.maxLength) {
+        Logger.warn("Attribute 'maxLength' is deprecated");
+      }
+      this.placeholder().html(componentHTML);
+    },
+
```

Comment:
I think that relying on jQuery's `val()` is safer. In any case, this function is more readable thus: 
```
function() {
  var attrs = { 
    type: "text",
    id: this.name,
    name: this.name,
    value: Dashboards.getParameterValue(this.parameter),
    size: this.size || this.charWidth || undefined,
    maxlength: this.maxLength || this.maxChars || undefined
};

var componentHTML = '<input ' + processAttrs(attrs) + '>';
if (this.addClearIcon){
  var iconAttrs = {
    id: this.name + "-clear-icon", 
    class: "clear-icon" + (this.clearIconClassName || "")
  };
  componentHTML += "<div " + processAttrs(iconAttrs) + "></div>";
}

if (this.size) Logger.warn("Attribute 'size' is deprecated");
if (this.maxLength) Logger.warn("Attribute 'maxLength' is deprecated");

this.placeholder().html(componentHTML);

function processAttrs(conf) {
  var list = [];
  for (prop in conf){
    if (conf.hasOwnProperty(prop). && conf[prop] != null){
      list.push(prop + '="' + Dashboards.escapeHtml(conf[prop]) + '"');
    }
  }
   return list.join(" ");
 }
}
```